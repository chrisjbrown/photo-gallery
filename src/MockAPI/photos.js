import data from "./data";

export const getPhotos = ({ page = 1, perPage = 10, query }) => {
  return new Promise(function (resolve) {
    const start = page === 1 ? 0 : (page - 1) * perPage;
    const end = page === 1 ? perPage : page * perPage;
    const photos = query
      ? data.photo.filter((photo) => photo.title.toLowerCase().includes(query))
      : data.photo;

    setTimeout(resolve, 300, {
      photos: photos.slice(start, end),
      pages: Math.ceil(photos.length / perPage),
    });
  });
};

export const patchPhoto = ({ photoPatch }) => {
  return new Promise(function (resolve) {
    const photoIndex = data.photo.findIndex(
      (photo) => photo.id === photoPatch.id
    );

    data.photo[photoIndex] = {
      ...data.photo[photoIndex],
      title: photoPatch.title,
      ispublic: photoPatch.isPublic,
      description: {
        _content: photoPatch.description,
      },
    };

    setTimeout(resolve, 300, data.photo[photoIndex]);
  });
};
