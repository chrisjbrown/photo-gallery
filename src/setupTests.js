import "@testing-library/jest-dom";
import { Router } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import "jest-styled-components";

export function renderWithRouter(
  ui,
  { history = createMemoryHistory(), ...renderOptions } = {}
) {
  function Wrapper({ children }) {
    return <Router history={history}>{children}</Router>;
  }

  return { ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
