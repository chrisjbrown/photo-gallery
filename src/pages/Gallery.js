import React, { useReducer, useState, useEffect } from "react";
import { useLocation, useHistory, useParams } from "react-router-dom";

import Divider from "../components/Divider";
import Button from "../components/Button";
import Modal from "../components/Modal";
import Pagination from "../components/Pagination";
import PhotoModal from "../components/PhotoModal";
import { getPhotos, patchPhoto } from "../MockAPI/photos.js";

import {
  Header,
  StyledLink,
  Controls,
  List,
  StyledImage,
} from "./Gallery.style";

const initialState = {
  pages: 1,
  photos: [],
};

const reducer = (state, action) => {
  switch (action.type) {
    case "SET_PHOTOS":
      return {
        ...state,
        photos: action.payload.photos,
        pages: action.payload.pages,
      };
    case "UPDATE_PHOTO": {
      const photos = state.photos.map((photo) => {
        if (photo.id === action.payload.photo.id) {
          return {
            ...photo,
            ...action.payload.photo,
          };
        }
        return photo;
      });
      return {
        ...state,
        photos,
      };
    }
    default:
      return state;
  }
};

function selectPhoto(state, id) {
  const photo = state.find((photo) => photo.id === id);

  if (photo) {
    return {
      id: photo.id,
      title: photo.title,
      description: photo.description._content,
      isPublic: photo.ispublic,
      ownerName: photo.ownername,
      dimensions: {
        width: photo.width_s,
        height: photo.height_s,
      },
    };
  }

  return null;
}

function Gallery() {
  const history = useHistory();
  const { photoId } = useParams();
  const searchParams = new URLSearchParams(useLocation().search);
  const page = Number.parseInt(searchParams.get("page")) || 1;
  const perPage = Number.parseInt(searchParams.get("perPage")) || 10;
  const query = searchParams.get("query") || "";

  const [selectedPhoto, setSelectedPhoto] = useState();
  const [isPendingPhotos, setPendingPhotos] = useState(true);
  const [{ pages, photos }, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    setSelectedPhoto(photoId ? selectPhoto(photos, photoId) : null);
  }, [photos, photoId]);

  useEffect(() => {
    setPendingPhotos(true);

    getPhotos({ page, perPage, query }).then((response) => {
      setPendingPhotos(false);
      dispatch({ type: "SET_PHOTOS", payload: response });
    });
  }, [page, perPage, query]);

  function updateSearch(e) {
    if (e.target.value) {
      searchParams.set("query", e.target.value);
    } else {
      searchParams.delete("query");
    }

    searchParams.set("page", 1);
    pushToHistory();
  }

  function updatePerPage(e) {
    searchParams.set("perPage", e.target.value);
    pushToHistory();
  }

  function updatePage(newPage) {
    if (newPage === 0 || newPage > pages) {
      return;
    }
    searchParams.set("page", newPage);
    pushToHistory();
  }

  function updatePhoto(photo) {
    patchPhoto({ photoPatch: photo }).then((response) => {
      dispatch({ type: "UPDATE_PHOTO", payload: { photo: response } });
    });
    closeModal();
  }

  function closeModal() {
    pushToHistory();
  }

  function pushToHistory() {
    history.push({ pathname: "/", search: `?${searchParams.toString()}` });
  }

  function renderPhotoModal() {
    if (selectedPhoto) {
      return (
        <PhotoModal
          photo={selectedPhoto}
          confirm={updatePhoto}
          cancel={closeModal}
        />
      );
    } else {
      const footer = (
        <React.Fragment>
          <Button onClick={closeModal}>Close</Button>
        </React.Fragment>
      );

      return (
        <Modal cancel={closeModal} footer={footer}>
          Oh no that photo wasn't found!
        </Modal>
      );
    }
  }

  const PhotoList = React.memo((props) => (
    <List pending={props.isPending}>
      {props.photos.map((photo) => (
        <StyledLink
          key={photo.id}
          to={{
            pathname: `/${photo.id}`,
            search: `?${searchParams.toString()}`,
          }}
        >
          <StyledImage
            title={photo.title}
            alt={photo.title}
            src={photo.url_s}
          />
        </StyledLink>
      ))}
    </List>
  ));

  return (
    <div>
      {photoId && renderPhotoModal()}
      <Header>Photo Gallery</Header>
      <Controls>
        <input name="search" placeholder="search" onChange={updateSearch} />
        <select value={perPage} onChange={updatePerPage}>
          <option value="10">10</option>
          <option value="20">20</option>
          <option value="30">30</option>
        </select>
        <div>
          Displaying page {page} of {pages}
        </div>
      </Controls>
      <Divider />
      <PhotoList isPending={isPendingPhotos} photos={photos} />
      <Divider />
      <Pagination page={page} pages={pages} updatePage={updatePage} />
    </div>
  );
}

export default Gallery;
