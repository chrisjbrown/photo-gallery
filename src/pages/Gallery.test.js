import React from "react";
import { fireEvent, act } from "@testing-library/react";
import { createMemoryHistory } from "history";

import { renderWithRouter } from "../setupTests";
import * as MockAPI from "../MockAPI/photos";
import Gallery from "./Gallery";

jest.mock("../MockAPI/photos", () => ({
  getPhotos: jest.fn(),
  patchPhoto: jest.fn(),
}));

describe("Gallery page", () => {
  beforeEach(() => {
    MockAPI.getPhotos.mockImplementation(({ perPage }) => {
      const photos = Array.apply(null, Array(perPage)).map((photo, index) => ({
        id: index,
        title: "tiki" + index,
      }));
      return Promise.resolve({
        photos,
        pages: Math.ceil(photos.length / perPage),
      });
    });
  });

  it("should have default view with no photos", async () => {
    let wrapper;
    MockAPI.getPhotos.mockImplementation((params) =>
      Promise.resolve({ photos: [] })
    );

    await act(async () => {
      wrapper = renderWithRouter(<Gallery />);
    });

    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  it("should have default view with 10 photos", async () => {
    let wrapper;

    await act(async () => {
      wrapper = renderWithRouter(<Gallery />);
    });

    const photos = wrapper.container.querySelectorAll("img");
    expect(photos.length).toEqual(10);
  });

  it("should call get photos with correct params from url and render response", async () => {
    let wrapper;
    const history = createMemoryHistory();
    history.push({ pathname: "/", search: "?page=2&perPage=20&query=tiki" });

    await act(async () => {
      wrapper = renderWithRouter(<Gallery />, { history });
    });

    expect(MockAPI.getPhotos).toHaveBeenCalledWith({
      page: 2,
      perPage: 20,
      query: "tiki",
    });

    const photos = wrapper.container.querySelectorAll("img");
    expect(photos.length).toEqual(20);
  });

  describe("User filtering", () => {
    it("should update based on query", async () => {
      let wrapper;

      await act(async () => {
        wrapper = renderWithRouter(<Gallery />);
      });

      const search = wrapper.getByRole("textbox");

      await act(async () => {
        fireEvent.input(search, { target: { value: "blabla" } });
      });

      expect(MockAPI.getPhotos).toHaveBeenCalledWith({
        page: 1,
        perPage: 10,
        query: "blabla",
      });
    });

    it("should update based on perPage", async () => {
      let wrapper;

      await act(async () => {
        wrapper = renderWithRouter(<Gallery />);
      });

      await act(async () => {
        fireEvent.change(wrapper.container.querySelector("select"), {
          target: { value: "20" },
        });
      });

      expect(MockAPI.getPhotos).toHaveBeenCalledWith({
        page: 1,
        perPage: 20,
        query: "",
      });
    });
  });

  describe("Pagination", () => {
    beforeEach(() => {
      MockAPI.getPhotos.mockImplementation(({ perPage }) => {
        const photos = Array.apply(null, Array(perPage * 3)).map(
          (photo, index) => ({
            id: index,
            title: "tiki" + index,
          })
        );
        return Promise.resolve({
          photos,
          pages: Math.ceil(photos.length / perPage),
        });
      });
    });

    it("should handle next/previous pagination", async () => {
      let wrapper;

      await act(async () => {
        wrapper = renderWithRouter(<Gallery />);
      });

      await act(async () => {
        fireEvent.click(wrapper.getByRole("button", { name: /next page/i }));
      });

      await act(async () => {
        fireEvent.click(wrapper.getByRole("button", { name: /next page/i }));
      });

      expect(MockAPI.getPhotos).toHaveBeenLastCalledWith({
        page: 3,
        perPage: 10,
        query: "",
      });

      await act(async () => {
        fireEvent.click(
          wrapper.getByRole("button", { name: /previous page/i })
        );
      });

      expect(MockAPI.getPhotos).toHaveBeenLastCalledWith({
        page: 2,
        perPage: 10,
        query: "",
      });
    });

    it("should handle end/start pagination", async () => {
      let wrapper;

      await act(async () => {
        wrapper = renderWithRouter(<Gallery />);
      });

      await act(async () => {
        fireEvent.click(wrapper.getByRole("button", { name: /last page/i }));
      });

      expect(MockAPI.getPhotos).toHaveBeenLastCalledWith({
        page: 3,
        perPage: 10,
        query: "",
      });

      await act(async () => {
        fireEvent.click(wrapper.getByRole("button", { name: /first page/i }));
      });

      expect(MockAPI.getPhotos).toHaveBeenLastCalledWith({
        page: 1,
        perPage: 10,
        query: "",
      });
    });
  });
});
