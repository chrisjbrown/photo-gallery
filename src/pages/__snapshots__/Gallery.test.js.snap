// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery page should have default view with no photos 1`] = `
.c2 {
  width: 100%;
  margin: 20px 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  padding: 10px 20px;
  outline: 0;
  cursor: pointer;
  border: 1px solid lightblue;
  pointer-events: none;
}

.c5:active {
  background-color: lightblue;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 20px;
}

.c0 {
  text-align: center;
}

.c1 {
  display: grid;
  grid-template-columns: repeat(auto-fit,200px);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 div:last-child {
  justify-self: flex-end;
}

.c3 {
  display: grid;
  grid-template-columns: repeat(auto-fit,240px);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-column-gap: 20px;
  column-gap: 20px;
  row-gap: 20px;
  min-height: 500px;
  opacity: 1;
}

<div>
  <h1
    class="c0"
  >
    Photo Gallery
  </h1>
  <div
    class="c1"
  >
    <input
      name="search"
      placeholder="search"
    />
    <select>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="20"
      >
        20
      </option>
      <option
        value="30"
      >
        30
      </option>
    </select>
    <div>
      Displaying page 
      1
       of 
    </div>
  </div>
  <hr
    class="c2"
  />
  <div
    class="c3"
  />
  <hr
    class="c2"
  />
  <div
    class="c4"
  >
    <button
      class="c5"
      disabled=""
      title="first page"
    >
      &lt;&lt;
    </button>
    <button
      class="c5"
      disabled=""
      title="previous page"
    >
      &lt;
    </button>
    <button
      class="c5"
      disabled=""
      title="next page"
    >
      &gt;
    </button>
    <button
      class="c5"
      disabled=""
      title="last page"
    >
      &gt;&gt;
    </button>
  </div>
</div>
`;
